plugins/package-managers/bazel/src/main/kotlin/CompositeBazelModuleRegistryService.kt

```diff
@@ -0,0 +1,102 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.plugins.packagemanagers.bazel
+
+import java.io.File
+
+import org.apache.logging.log4j.kotlin.logger
+
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.BazelModuleRegistryService
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.LocalBazelModuleRegistryService
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.ModuleMetadata
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.ModuleSourceInfo
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.RemoteBazelModuleRegistryService
+
+/**
+ * A composite Bazel module registry service that aggregates multiple [BazelModuleRegistryService] instances and
+ * delegates the package lookup to the corresponding registry. The first registry that contains the package is used.
+ * [packagesPerRegistry] is a map of [BazelModuleRegistryService] instances to the list of packages they are
+ * responsible for.
+ */
+class CompositeBazelModuleRegistryService(
+    private val packagesPerRegistry: Map<BazelModuleRegistryService, List<String>>
+) : BazelModuleRegistryService {
+    companion object {
+        // A regular expression to extract the server and package parts of a registry URL.
+        internal val URL_REGEX = "^(?<server>.*/)modules/(?<package>[^/]+)/[^/]+/source\\.json$".toRegex()
+
+        /**
+         * Create a Composite Bazel Module Registry client instance.
+         */
+        fun create(urls: Set<String>, projectDir: File): CompositeBazelModuleRegistryService {
+            val unknownKey = "UNKNOWN"
+
+            val packagesPerServer = urls.filter { it.endsWith("source.json") }.map {
+                val groups = URL_REGEX.matchEntire(it)?.groups
+
+                val key = groups?.get("server")?.value ?: unknownKey
+                val value = groups?.get("package")?.value ?: it
+
+                key to value
+            }.groupByTo(mutableMapOf(), { it.first }) { it.second }
+
+            packagesPerServer.remove(unknownKey)?.let {
+                logger.warn {
+                    "The following URLs cannot be mapped to a server root:"
+                }
+
+                it.forEach { logger.warn { it } }
+            }
+
+            val packagesPerRegistry = packagesPerServer.mapKeys { entry ->
+                val url = entry.key
+
+                if (url.startsWith("file://")) {
+                    val localRegistryURL = url.replace("%workspace%", projectDir.absolutePath)
```

Comment:
I believe `%workspace%` should be responsibility of client code. It just seems closer related there. This way this class can be tested more easily, because it also would not know about the project dir.

Reply:
As you mentioned yesterday, the client code does not currently know about the project dir.
The `LocalBazelModuleRegistryService` currently expects a`File`.
Are you suggesting changing it to a `URL` just to be able to do the `%workspace%` expansion?

Reply:
I suggest that the code which calls the `create()` function replaces `%workspace%` in the URLs.